docs/proposals/templates.md

```diff
@@ -0,0 +1,350 @@
+template objects.  However since there is a desire to at least be able to support a server side implementation, this proposal
+does assume the specification will be k8s API friendly.
+
+## Desired characteristics
+
+* K8S object json-compliant syntax.  This allows server side apis that align with existing k8s apis to be constructed 
+  which consume templates and existing k8s tooling to work with them.  It also allows for api versioning/migration to be managed by
+  the existing k8s codec scheme rather than having to define/introduce a new syntax evolution mechanism.
+  * (Even if they are not part of the k8s core, it would still be good if a server side template processing+managing api supplied 
+    as an ApiGroup consumed the same k8s object schema as the peer k8s apis rather than introducing a new one)
+  * This goal would be met by a solution which wraps some other template content (eg jinja, python, etc) inside a k8s object wrapper
+* Self-contained parameter definitions.  This allows a template to be a portable object which includes metadata that describe
+  the inputs it expects, making it easy to wrapper a user interface around the parameterization flow.
+* Support for non-string parameters.  K8S objects define fields of type string, int, bool, and byte[](base64).  The parameter
+  scheme should support all of those types for substitution.
+  * complex types (struct/map/list) can be defined in terms of the available primitives, so it's preferred to avoid the complexity
+    of allowing for full complex-type substitution.
+* Parameter metadata.  Parameters should include at a minimum, information describing the purpose of the parameter, whether it is 
+  required/optional, and a default/suggested value.  Type information could also be required to resolve the non-string substitution
+  requirement.
+* Parameter type support:  should support substituting string/int/bool and base64 encoded byte[] values.  maps/lists/structs can
+  be built up manually.
+* Template metadata.  Templates should be able to include metadata describing their purpose or links to further documentation and 
+  versioning information.
+* Pluggable template expansion.  It should be possible to define multiple template expansion strategies and the system should
+  delegate a template to the correct plugin for expansion and then return the result.
+
+## Proposed Object Syntax
+
+Example of a template definition:
+```
+kind: TemplateDeclaration
+apiVersion: v1
+metadata: 
+  name: myHelloWorldTemplate
+templateKind: python
+params:
+  - name: message
+    required: true
+    default_value: Hello World!
+    type: string
+  - name: port
+    required: true
+    default_value: 8080
+    type: int
+includes:
+  - path: https://github.com/myorg/mytemplates/template1.yml
+  - path: https://github.com/myorg/mytemplates/template2.yml
+body: |
+  import json
+  import kubernetes_tmpl
+  # Real templates would be doing actual computation.
+  print json.dumps({
```

Comment:
let's not worry about the specific content of the envelope, the purpose of introducing the generic "body" concept was so we didn't have to worry about what the body looks like since multiple plugins can be implemented to support different syntaxes.